### PR TITLE
DEV-27 allow non-indexd property updates even when file is validated

### PR DESCRIPTION
**Pull request by anmaxvl**
Previous behavior did not allow any updates to file type nodes once
file in S3 has been validated. This is not ideal, since this also
disables updating metadata that is not file related, e.g.
read_pair_number on a SAR. This PR relaxes this constraint and allows
updates when no changes to IndexD records are made. Also extend
"updatable" file states with "deleted" state.

### CODE REVIEW

File: sheepdog/transactions/upload/sub_entities.py [hunk 1]
```diff
@@ -3,18 +3,18 @@
...
 from indexclient.client import Document
 from sqlalchemy.orm.exc import NoResultFound
-from sheepdog.globals import PRIMARY_URL_TYPE
 
 from sheepdog.errors import UserError
 from sheepdog.globals import (
     DATA_FILE_CATEGORIES, PRIMARY_URL_TYPE, POSSIBLE_OPEN_FILE_NODES,
     UPDATABLE_FILE_STATES, RELEASED_NODE_STATES,
-    MODIFIABLE_RELEASED_FILE_STATES
+    MODIFIABLE_RELEASED_FILE_STATES, INDEXD_METADATA_FIELDS,
```
Comment (kulgan) [inline sheepdog/transactions/upload/sub_entities.py:17]: don't think this was used

File: sheepdog/transactions/upload/sub_entities.py [hunk 2]
```diff
@@ -473,22 +480,52 @@ def is_updatable_file_node(self, node):
...
-                 "this data.").format(file_state),
-                keys=['file_state'],
-                type=EntityErrors.INVALID_PERMISSIONS,
-            )
+        if file_state in UPDATABLE_FILE_STATES:
+            return True
+
+        # File state is "validated" and we need to make sure that we're not
+        # changing any IndexD metadata
+        if self._is_identical_indexd_metadata():
```
Comment (scintillating7) [inline sheepdog/transactions/upload/sub_entities.py:497]: I appreciate your abstraction layers 👍

File: sheepdog/transactions/upload/sub_entities.py [hunk 3]
```diff
@@ -766,7 +806,7 @@ def _is_modifiable_release(self, node):
             return current_release is not None and file_state in MODIFIABLE_RELEASED_FILE_STATES
         return False
 
-    def get_indexed_document(self):
+    def get_indexed_document(self) -> Optional[Document]:
```
Comment (scintillating7) [inline sheepdog/transactions/upload/sub_entities.py:809]: this is a nice change. do you want to add type hints for the other functions?

File: tests/integration/submission/conftest.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def patch_envelopes(smtpserver, monkeypatch):
     def mock_smtp(*args, **kwargs):
         """Monkeypatch SMTP to return a connection to test server"""
 
-        host, port = smtpserver.addr
+        host, port = smtpserver.addr[:2]
```
Comment (scintillating7) [inline tests/integration/submission/conftest.py:189]: can this have more than the host and port in it, then?

File: sheepdog/transactions/upload/sub_entities.py (hunk 3 above)
Comment (anmaxvl) [reply sheepdog/transactions/upload/sub_entities.py:809]: yeah, sure. but I'll limit it to the scope of this PR 😄

File: tests/integration/submission/conftest.py (hunk 4 above)
Comment (anmaxvl) [reply tests/integration/submission/conftest.py:189]: yeah, maybe it's only in the version that was pulled in locally, but this is safer regardless

File: sheepdog/transactions/upload/sub_entities.py [hunk 5]
```diff
@@ -473,22 +481,60 @@ def is_updatable_file_node(self, node):
         Return:
             bool: whether the node is an updatable file
         """
-        if not node._dictionary.get('category') in DATA_FILE_CATEGORIES:
-            return False
+
+        if not is_data_file_node(node):
+            return True
```
[the remark this answers is not in the document]
Comment (anmaxvl) [reply sheepdog/transactions/upload/sub_entities.py:486]: ideally we shouldn't even have this ever evaluate to `True`, I'll remove it

File: sheepdog/transactions/upload/sub_entities.py [hunk 6]
```diff
@@ -473,22 +481,60 @@ def is_updatable_file_node(self, node):
...
+
+        Check if the following properties haven't changed:
+            * file_size -> Document.size
+            * file_name -> Document.file_name
+            * md5sum -> Document.hashes["md5"]
+
+        """
+        doc = self.get_indexed_document()
+
+        file_size = self.doc.get("file_size")
```
[the remark this answers is not in the document]
Comment (anmaxvl) [reply sheepdog/transactions/upload/sub_entities.py:526]: 👍

**Pull request merged**